### PR TITLE
feat: support downloading project artifacts for the new bucket layout

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
It exposes a new method: `FetchProjectBinary` so that the existing `FetchBeatsBinary` calls the new one with the required arguments.

This method will use the new bucket layout:

- $project
  - commits
    -  SHA1
      - binary
  - snapshots
    - binary-version

And it will look up the artifacts in the three possible locations:
1. the new project layout (as shown above) in the fleet-ci bucket
2. the new project layout (as shown above) in the beats-ci bucket
3. the new Beats layout (same as legacy but adding a `beats` parent dir)
4. the legacy Beats layout

This PR is decoupling the bucket resolution into the concept of URL Resolvers, so that the logic is narrowed to the scope of that struct. We are building three resolvers, one for each layout. We have refactored the existing code that calculated the bucket URL using the new resolvers abstraction, checking that the existing unit tests covered the changes. Finally, we moved those tests to the right Go file, so that it's easier to maintain them.

It's important to mention that we did not change the API contracts in the pkg package: we are aware of they are consumed (by elastic-agent i.e.) and we are keeping old functionality using defaults. The framework is using new methods that support the new expected behaviors:
- selectively use the CI snapshots from the bucket for downloads: when installing elastic-agent on top of beats, we need to install metricbeat and filebeat first. Their versions are going to be aligned with the stack, so we need to fetch them from the Elastic artifactory.
- fetch CI artifacts for a project, in an exhaustive manner: we will try all possible locations to fetch the CI snapshots, being possible to use the Fleet CI bucket or the Beats CI one, in the new and legacy layout.

Apart from that, we realised that the VM support did not migrate the ability to consume a Git SHA from Beats. In this PR we are adding back that support, simply passing the env vars defined at the Jenkinsfile to the Ansible playbook to generate the right provisioning script.

## Why is it important?
Both fleet-server and elastic-agent projects will follow this structure, and because we want to consume the artifacts for PRs, then we need to update the method to look up the bucket.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have run the Unit tests (`make unit-test`), and they are passing locally
- [ ] I have run the End-2-End tests for the suite I'm working on, and they are passing locally
- [ ] I have noticed new Go dependencies (run `make notice` in the proper directory)

## Related issues

- Closes https://github.com/elastic/e2e-testing/issues/2149

## Other concerns
We are aware that passing the GIT_SHA of the elastic-agent could cause errors in the k8s-autodiscover test suite, because they download the binaries for metricbeat and filebeat in that commit. Because the SHA for the elastic-agent comes from separate repository, the binaries for MB and FB don't exist for that commit. In https://github.com/elastic/e2e-testing/issues/2175 we explain it more in depth

### CODE REVIEW

**Review by narph — APPROVED** [review]
LGTM

File: pkg/downloads/versions.go [hunk 1]
```diff
@@ -434,9 +445,22 @@ func FetchBeatsBinary(ctx context.Context, artifactName string, artifact string,
...
-		bucket, prefix, object := getGCPBucketCoordinates(artifactName, artifact)
+		variant := ""
+		if strings.HasSuffix(artifactName, "-ubi8") {
+			variant = "ubi8"
+		}
+
+		// look up the bucket in this particular order:
+		// 1. the project layout (elastic-agent, fleet-server)
+		// 2. the new beats layout (beats)
+		// 3. the legacy Beats layout (commits/snapshots)
```
Comment (adam-stokes) [inline pkg/downloads/versions.go:456]: very nice

**Review by adam-stokes — APPROVED** [review]
Looks awesome! 🥳

**Comment by mdelapenya** [comment]
Need to run it for a commit Sha, will try later on

File: .ci/ansible/tasks/setup_test_script.yml [hunk 2]
```diff
@@ -85,6 +92,7 @@
   with_items:
     - "export STACK_VERSION={{ lookup('file', '{{workspace}}/.stack-version') or '8.0.0-SNAPSHOT' }}"
     - "export BEAT_VERSION={{ lookup('file', '{{workspace}}/.stack-version') or '8.0.0-SNAPSHOT' }}"
+    - "export GITHUB_CHECK_SHA1=={{ lookup('env', 'GITHUB_CHECK_SHA1') }}"
```
Comment (mdelapenya) [inline .ci/ansible/tasks/setup_test_script.yml:95]:
```suggestion
    - "export GITHUB_CHECK_SHA1={{ lookup('env', 'GITHUB_CHECK_SHA1') }}"
```

File: .ci/ansible/tasks/setup_test_script.yml [hunk 3]
```diff
@@ -92,7 +92,7 @@
   with_items:
     - "export STACK_VERSION={{ lookup('file', '{{workspace}}/.stack-version') or '8.0.0-SNAPSHOT' }}"
     - "export BEAT_VERSION={{ lookup('file', '{{workspace}}/.stack-version') or '8.0.0-SNAPSHOT' }}"
-    - "export GITHUB_CHECK_SHA1=={{ lookup('env', 'GITHUB_CHECK_SHA1') }}"
+    - "export GITHUB_CHECK_SHA1={{ lookup('env', 'GITHUB_CHECK_SHA1') }}"
```
Comment (adam-stokes) [inline .ci/ansible/tasks/setup_test_script.yml:95]: Doh typos

Comment (mdelapenya) [reply .ci/ansible/tasks/setup_test_script.yml:95]: Fixed!

**Comment by mdelapenya** [comment]
I'm trying with different combination of commits:

- `b7349792853152ee195a8740b4c7c3bc0afae951`: which represent a commit in the new elastic-agent repository. The binaries should live in the new layout, and indeed this PR fetches from there, using the new ProjectURL resolver. But that commit contains an error that @narph already resolved. Trying with [`be6fc68886e1e11c74ff927a04192e3c674d7e01`](https://github.com/elastic/elastic-agent/commit/be6fc68886e1e11c74ff927a04192e3c674d7e01) in https://beats-ci.elastic.co/job/e2e-tests/job/e2e-testing-mbp/job/PR-2172/15/ instead

- `b1d41b1a37ec466f457e520c7e544abc8e596dae` which is current HEAD at Beats repo, having the CI already generated the binaries for that commit. CI pipeline: https://beats-ci.elastic.co/job/e2e-tests/job/e2e-testing-mbp/job/PR-2172/14/. This commit should use the last resolver, the BeatsLegacyURL resolver.

>[2022-02-24T11:02:21.352Z] [37mDEBU[0m[2022-02-24T11:02:21Z] Resolving URL from Project resolver           [37mbucket[0m=beats-ci-artifacts [37mobject[0m=elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [37mprefix[0m=elastic-agent/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae [37mproject[0m=elastic-agent
[2022-02-24T11:02:21.352Z] [37mTRAC[0m[2022-02-24T11:02:21Z] Executing request                             [37mescapedURL[0m="https://storage.googleapis.com/storage/v1/b/beats-ci-artifacts/o?prefix=elastic-agent/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae" [37mmethod[0m=GET
[2022-02-24T11:02:21.353Z] [37mTRAC[0m[2022-02-24T11:02:21Z] Google Cloud Storage API is available         [37mbucket[0m=beats-ci-artifacts [37melapsedTime[0m=98.625036ms [37mobject[0m=elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [37mprefix[0m=elastic-agent/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae [37mretries[0m=1 [37murl[0m="https://storage.googleapis.com/storage/v1/b/beats-ci-artifacts/o?prefix=elastic-agent/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae"
[2022-02-24T11:02:21.353Z] [37mDEBU[0m[2022-02-24T11:02:21Z] Objects found                                 [37mbucket[0m=beats-ci-artifacts [37mobject[0m=elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [37mobjects[0m=0 [37mprefix[0m=elastic-agent/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae
[2022-02-24T11:02:21.353Z] [33mWARN[0m[2022-02-24T11:02:21Z] the elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz object could not be found in the current page (0) in the beats-ci-artifacts bucket and elastic-agent/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae prefix  [33mbucket[0m=beats-ci-artifacts [33mcurrentPage[0m=0 [33mobject[0m=elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [33mprefix[0m=elastic-agent/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae
[2022-02-24T11:02:21.353Z] [33mWARN[0m[2022-02-24T11:02:21Z] Reached the end of the pages and the object was not found  [33mbucket[0m=beats-ci-artifacts [33mcurrentPage[0m=0 [33mobject[0m=elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [33mprefix[0m=elastic-agent/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae
[2022-02-24T11:02:21.353Z] [33mWARN[0m[2022-02-24T11:02:21Z] Object not found. Trying with another artifact resolver  [33mresolver[0m="&{beats-ci-artifacts elastic-agent elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz}"
[2022-02-24T11:02:21.353Z] [37mDEBU[0m[2022-02-24T11:02:21Z] Resolving URL from Beats resolver             [37mbucket[0m=beats-ci-artifacts [37mobject[0m=elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [37mprefix[0m=beats/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae [37mproject[0m=elastic-agent [37mvariant[0m=
[2022-02-24T11:02:21.353Z] [37mTRAC[0m[2022-02-24T11:02:21Z] Executing request                             [37mescapedURL[0m="https://storage.googleapis.com/storage/v1/b/beats-ci-artifacts/o?prefix=beats/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae" [37mmethod[0m=GET
[2022-02-24T11:02:21.353Z] [37mTRAC[0m[2022-02-24T11:02:21Z] Google Cloud Storage API is available         [37mbucket[0m=beats-ci-artifacts [37melapsedTime[0m=45.199685ms [37mobject[0m=elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [37mprefix[0m=beats/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae [37mretries[0m=1 [37murl[0m="https://storage.googleapis.com/storage/v1/b/beats-ci-artifacts/o?prefix=beats/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae"
[2022-02-24T11:02:21.353Z] [37mDEBU[0m[2022-02-24T11:02:21Z] Objects found                                 [37mbucket[0m=beats-ci-artifacts [37mobject[0m=elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [37mobjects[0m=0 [37mprefix[0m=beats/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae
[2022-02-24T11:02:21.354Z] [33mWARN[0m[2022-02-24T11:02:21Z] the elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz object could not be found in the current page (0) in the beats-ci-artifacts bucket and beats/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae prefix  [33mbucket[0m=beats-ci-artifacts [33mcurrentPage[0m=0 [33mobject[0m=elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [33mprefix[0m=beats/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae
[2022-02-24T11:02:21.354Z] [33mWARN[0m[2022-02-24T11:02:21Z] Reached the end of the pages and the object was not found  [33mbucket[0m=beats-ci-artifacts [33mcurrentPage[0m=0 [33mobject[0m=elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [33mprefix[0m=beats/commits/b1d41b1a37ec466f457e520c7e544abc8e596dae
[2022-02-24T11:02:21.354Z] [33mWARN[0m[2022-02-24T11:02:21Z] Object not found. Trying with another artifact resolver  [33mresolver[0m="&{beats-ci-artifacts elastic-agent  elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz}"
[2022-02-24T11:02:21.354Z] [37mDEBU[0m[2022-02-24T11:02:21Z] Resolving URL from Beats Legacy resolver      [37mbucket[0m=beats-ci-artifacts [37mobject[0m=elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [37mprefix[0m=commits/b1d41b1a37ec466f457e520c7e544abc8e596dae [37mproject[0m=elastic-agent [37mvariant[0m=
[2022-02-24T11:02:21.354Z] [37mTRAC[0m[2022-02-24T11:02:21Z] Executing request                             [37mescapedURL[0m="https://storage.googleapis.com/storage/v1/b/beats-ci-artifacts/o?prefix=commits/b1d41b1a37ec466f457e520c7e544abc8e596dae" [37mmethod[0m=GET
[2022-02-24T11:02:21.613Z] [37mTRAC[0m[2022-02-24T11:02:21Z] Google Cloud Storage API is available         [37mbucket[0m=beats-ci-artifacts [37melapsedTime[0m=133.895936ms [37mobject[0m=elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [37mprefix[0m=commits/b1d41b1a37ec466f457e520c7e544abc8e596dae [37mretries[0m=1 [37murl[0m="https://storage.googleapis.com/storage/v1/b/beats-ci-artifacts/o?prefix=commits/b1d41b1a37ec466f457e520c7e544abc8e596dae"
[2022-02-24T11:02:21.613Z] [37mDEBU[0m[2022-02-24T11:02:21Z] Objects found                                 [37mbucket[0m=beats-ci-artifacts [37mobject[0m=elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [37mobjects[0m=363 [37mprefix[0m=commits/b1d41b1a37ec466f457e520c7e544abc8e596dae
[2022-02-24T11:02:21.613Z] [36mINFO[0m[2022-02-24T11:02:21Z] medialink: https://storage.googleapis.com/download/storage/v1/b/beats-ci-artifacts/o/commits%2Fb1d41b1a37ec466f457e520c7e544abc8e596dae%2Felastic-agent%2Felastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz?generation=1645674571811401&alt=media 
[2022-02-24T11:02:21.613Z] [37mDEBU[0m[2022-02-24T11:02:21Z] Media link found for the object               [37mbucket[0m=beats-ci-artifacts [37melapsedTime[0m=137.984462ms [37mmedialink[0m="https://storage.googleapis.com/download/storage/v1/b/beats-ci-artifacts/o/commits%2Fb1d41b1a37ec466f457e520c7e544abc8e596dae%2Felastic-agent%2Felastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz?generation=1645674571811401&alt=media" [37mobject[0m=elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz [37mprefix[0m=commits/b1d41b1a37ec466f457e520c7e544abc8e596dae [37mretries[0m=1
[2022-02-24T11:02:21.613Z] [37mTRAC[0m[2022-02-24T11:02:21Z] Downloading file                              [37mpath[0m=/tmp/1e555740-8032-4178-81f5-5382a67aeb08/9662bbf9-c4ee-4b52-b87b-9367ea12b562 [37murl[0m="https://storage.googleapis.com/download/storage/v1/b/beats-ci-artifacts/o/commits%2Fb1d41b1a37ec466f457e520c7e544abc8e596dae%2Felastic-agent%2Felastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz?generation=1645674571811401&alt=media"
[2022-02-24T11:02:21.614Z] [37mTRAC[0m[2022-02-24T11:02:21Z] File downloaded                               [37melapsedTime[0m=62.682719ms [37mpath[0m=/tmp/1e555740-8032-4178-81f5-5382a67aeb08/9662bbf9-c4ee-4b52-b87b-9367ea12b562 [37mretries[0m=1 [37murl[0m="https://storage.googleapis.com/download/storage/v1/b/beats-ci-artifacts/o/commits%2Fb1d41b1a37ec466f457e520c7e544abc8e596dae%2Felastic-agent%2Felastic-agent-8.2.0-SNAPSHOT-linux-x86_64.tar.gz?generation=1645674571811401&alt=media"

- `20ec08851cc472b439229b97578e63aa45ef9cbc` which is a commit at Beats repo, having the CI already generated the binaries for the commit using the new layout. Therefore, the new BeatsURL resolver should retrieve it.

File: internal/installer/elasticagent_deb.go [hunk 4]
```diff
@@ -170,15 +170,15 @@ func (i *elasticAgentDEBPackage) Preinstall(ctx context.Context) error {
 
 	for _, bp := range i.service.BackgroundProcesses {
 		if strings.EqualFold(bp, "filebeat") || strings.EqualFold(bp, "metricbeat") {
-			// pre-install the dependant binary first
-			err := installArtifactFn(ctx, bp)
+			// pre-install the dependant binary first, using the stack version
+			err := installArtifactFn(ctx, bp, common.StackVersion)
```
Comment (mdelapenya) [inline internal/installer/elasticagent_deb.go:174]: Should we use the current stack version of the current released version?

**Comment by mdelapenya** [comment]
@adam-stokes I think this is ready to merge. Please do a final review 👀 with latest changes 🙏 

The 13 errors are related to k8s-autodiscover **for a SHA** commit: because the last build used a SHA on elastic-agent repo, the binaries used for the k8s feature are not found by any resolver, as the SHA commit does not exist in the Beats repository (and its bucket). In https://github.com/elastic/e2e-testing/issues/2175 I explain the need to refactor the layout of the test suites.

**Comment by mdelapenya** [comment]
### Tests for a commit on Beats
- Commit: [37d07d1cd3ea761939fbe5f568b747511b302c0c](https://github.com/elastic/beats/commit/37d07d1cd3ea761939fbe5f568b747511b302c0c)
- CI Build: https://beats-ci.elastic.co/blue/organizations/jenkins/e2e-tests%2Fe2e-testing-mbp/detail/PR-2172/27/pipeline/485
- Bucket: `beats-ci-artifacts/commits/37d07d1cd3ea761939fbe5f568b747511b302c0c` 
- Should use elastic-agent in STACK_VERSION

### Tests for a commit on Elastic Agent
- Commit: [e921876ddd33cf6053b1487d7b32e17965b79159](https://github.com/elastic/elastic-agent/commit/e921876ddd33cf6053b1487d7b32e17965b79159)
- CI Build: 
- Bucket: `fleet-ci-artifacts/elastic-agent/commits/e921876ddd33cf6053b1487d7b32e17965b79159`
- Should use beats (filebeat, metricbeat) in STACK_VERSION

**Review by adam-stokes — APPROVED** [review]

File: pkg/downloads/buckets.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+// Resolve returns the bucket, prefix and object for Beats artifacts
+func (r *BeatsURLResolver) Resolve() (string, string, string) {
+	artifact := r.Beat
+	fileName := r.FileName
+
+	if strings.EqualFold(r.Variant, "ubi8") {
+		artifact = strings.ReplaceAll(artifact, "-ubi8", "")
+	}
```
Comment (mdelapenya) [inline pkg/downloads/buckets.go:107]: Candidate to be extracted, as it's used across all resolvers

File: pkg/downloads/buckets.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+	prefix := fmt.Sprintf("beats/snapshots/%s", artifact)
+	object := fileName
+
+	ciSnapshotsFn := UseBeatsCISnapshots
+	if strings.EqualFold(artifact, "elastic-agent") {
+		ciSnapshotsFn = UseElasticAgentCISnapshots
+	}
```
Comment (mdelapenya) [inline pkg/downloads/buckets.go:115]: Candidate to be extracted, as it's used across all resolvers

**Pull request merged**